server/libs/storage/src/storage.service.ts

```diff
@@ -0,0 +1,110 @@
+import { APP_UPLOAD_LOCATION } from '@app/common';
+import { AssetEntity } from '@app/database/entities/asset.entity';
+import { ImmichConfigService } from '@app/immich-config';
+import { Injectable, Logger } from '@nestjs/common';
+import { InjectRepository } from '@nestjs/typeorm';
+import { Repository } from 'typeorm';
+import { FileSystemStorageService } from './storage.service.filesystem';
+import handlebar from 'handlebars';
+import * as luxon from 'luxon';
+import sanitize from 'sanitize-filename';
+import path from 'node:path';
+import { constants } from 'node:fs';
+import fsPromise from 'fs/promises';
+import mv from 'mv';
+
+@Injectable()
+export class StorageService {
+  readonly log = new Logger(StorageService.name);
+
+  constructor(
+    @InjectRepository(AssetEntity)
+    private assetRepository: Repository<AssetEntity>,
+
+    private fileSystemStorageService: FileSystemStorageService,
+
+    private immichConfigService: ImmichConfigService,
+  ) {}
+
+  public async moveFile(asset: AssetEntity, filename: string): Promise<AssetEntity> {
+    try {
+      const parsedPath = await this.buildPath(asset, filename);
+      let duplicateCount = 1;
+      let qualifiedPath = path.join(APP_UPLOAD_LOCATION, asset.userId, parsedPath);
+
+      while (await this.checkFileExist(qualifiedPath)) {
+        const newPath = await this.buildPath(asset, filename + `_${duplicateCount}`);
+        qualifiedPath = path.join(APP_UPLOAD_LOCATION, asset.userId, newPath);
+        duplicateCount++;
+      }
+
+      const destination = await this.moveAndDeleteFile(asset.originalPath, qualifiedPath);
+
+      asset.originalPath = destination;
+      return await this.assetRepository.save(asset);
+    } catch (e) {
+      this.log.error(e);
+      return asset;
+    }
+  }
+
+  private moveAndDeleteFile(original: string, destination: string): Promise<string> {
+    return new Promise((resolve, reject) => {
+      mv(original, destination, { mkdirp: true, clobber: false }, (err) => {
+        if (err) {
+          reject(err);
+        }
+
+        resolve(destination);
+      });
+    });
+  }
+
+  private async checkFileExist(path: string): Promise<boolean> {
+    try {
+      await fsPromise.access(path, constants.F_OK);
+      return true;
+    } catch (e) {
+      return false;
+    }
+  }
+
+  private async buildPath(asset: AssetEntity, filename: string) {
+    // Build path from user config
+    // Get user config from database
+    const configs = await this.immichConfigService.getConfig();
+
+    try {
+      const template = handlebar.compile(configs.storageTemplate.template, {
+        knownHelpers: undefined,
+      });
+
+      const dt = luxon.DateTime.fromISO(new Date(asset.createdAt).toISOString());
+      const sanitizedFilename = sanitize(path.basename(filename, path.extname(filename)));
+      const fileExtension = path.extname(asset.originalPath).split('.').pop();
```

Comment:
What is the purpose of the split and pop? I don't think you need these if you are using `path.extname`.

Reply:
`path.extname()` would return `.extension`, so this action is just to get the `extension` part

Reply:
Ah, to remove the leading dot, I see.

Reply:
Removed this implementation on getting rid of `{{ext}}`